vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_context.h

```diff
@@ -0,0 +1,272 @@
+    LogStream& operator=(LogStream&& other);
+
+    ~LogStream();
+
+    template <typename T>
+    LogStream& operator<<(const T& value) {
+      stream_ << value;
+      return *this;
+    }
+
+   private:
+    BATLedgerContext* context_;
+    base::Location location_;
+    LogLevel log_level_;
+    std::ostringstream stream_;
+    bool moved_ = false;
+  };
+
+  explicit BATLedgerContext(LedgerClient* ledger_client);
+  explicit BATLedgerContext(LedgerImpl* ledger_impl);
+
+  ~BATLedgerContext();
+
+  BATLedgerContext(const BATLedgerContext&) = delete;
+  BATLedgerContext& operator=(const BATLedgerContext&) = delete;
+
+  // Returns a pointer to the LedgerClient associated with this ledger context.
+  // In general, this method should only be used by low-level components that
+  // interact directly with the client. Before using this function in an
+  // application-level component, check to see if there is an existing component
+  // that wraps the low-level calls, and if not, consider creating one.
+  LedgerClient* GetLedgerClient() const { return ledger_client_; }
+
+  // Returns a pointer to the LedgerImpl associated with this ledger context.
+  // This method is provided for backward-compatibility with existing code that
+  // does not expose a context component and will be removed in the future.
+  LedgerImpl* GetLedgerImpl() const { return ledger_impl_; }
+
+  // Returns a pointer to the singleton context Component of type T. T must
+  // expose a static |kComponentKey| member. The singleton will be created if
+  // necessary.
+  //
+  // Example:
+  //   auto* my_component = context()->Get<MyComponent>();
+  template <typename T>
+  T* Get() {
+    static_assert(std::is_base_of<Component, T>::value,
+                  "Get<T> requires that T is a subclass of Component");
+    size_t key = T::kComponentKey.value_;
+    auto iter = components_.find(key);
+    if (iter != components_.end())
+      return reinterpret_cast<T*>(iter->second.get());
+
+    std::unique_ptr<T> instance(new T(this));
+    T* ptr = instance.get();
+    components_[key] = std::move(instance);
+    return ptr;
+  }
+
+  template <typename T>
+  void SetComponentForTesting(std::unique_ptr<T> component) {
+    static_assert(std::is_base_of<Component, T>::value,
+                  "SetComponentForTesting<T> requires that T is a subclass of "
+                  "Component");
+    size_t key = T::kComponentKey.value_;
+    components_[key] = std::move(component);
+  }
+
+  // Starts a component task and returns the AsyncResult associated with the
+  // task. When the task completes, the component is destroyed.
+  //
+  // Example:
+  //   auto result = context()->StartTask<MyTask>("hello");
+  //   result.Listen(...);
+  template <typename T, typename... Args>
+  auto StartTask(Args&&... args) {
+    std::unique_ptr<T> instance(new T(this));
+    T* ptr = instance.get();
+
+    size_t key = std::hash<T*>()(ptr);
+    tasks_[key] = std::move(instance);
```

Comment:
btw why this is not just a base::flat_set ?

Reply:
Yes, you're right, it should be. Or just `std::set`, but the number of entries (i.e. concurrent tasks) will likely be smallish so a `flat_set` is probably best.

Reply:
Oh, you're also asking why a map? I wanted an `std::unique_ptr` in the collection and I think I ran into some minor difficulty using that in a set, but I forget at the moment.

Reply:
yeah you won't need that juggling with keys when just using a set